File: folio/generator/xref.py
from __future__ import annotations

import re

from folio.ir import ClassIR, ModuleIR

# Python builtins that should never be linked
_BUILTINS = frozenset(
    {
        "str",
        "int",
        "float",
        "bool",
        "bytes",
        "bytearray",
        "list",
        "dict",
        "set",
        "frozenset",
        "tuple",
        "None",
        "type",
        "object",
        "complex",
        "range",
        "memoryview",
        "slice",
        "property",
        "classmethod",
        "staticmethod",
        "super",
        "Exception",
        "BaseException",
        "Any",
        "Callable",
        "Iterator",
        "Generator",
        "Coroutine",
        "Awaitable",
        "AsyncIterator",
        "AsyncGenerator",
        "Iterable",
        "Sequence",
        "Mapping",
        "MutableMapping",
        "MutableSequence",
        "MutableSet",
        "Optional",
        "Union",
        "Type",
        "ClassVar",
        "Final",
        "Literal",
        "Protocol",
        "TypeVar",
        "TypeAlias",
        "Self",
        "Never",
        "NoReturn",
        "Concatenate",
        "ParamSpec",
        "TypeVarTuple",
        "Unpack",
    }
)

# Regex to strip generic wrappers and union types, extracting inner type names
_GENERIC_RE = re.compile(r"^(\w+)\[(.+)\]$")
_UNION_PIPE_RE = re.compile(r"\s*\|\s*")


def build_symbol_index(
    modules: list[ModuleIR],
    docs_route_base: str = "/docs",
) -> dict[str, str]:
    """Build a mapping of fully-qualified symbol names to documentation URLs.

    Returns a dict like:
        {
            "folio.config": "/docs/api-reference/folio/config",
            "folio.config.Config": "/docs/api-reference/folio/config#config",
            "folio.config.load_config": "/docs/api-reference/folio/config#load_config",
        }
    """
    index: dict[str, str] = {}
    docs_route_base = docs_route_base.rstrip("/") or "/docs"

    for mod in modules:
        mod_route = f"{docs_route_base}/api-reference/{mod.name.replace('.', '/')}"
        index[mod.name] = mod_route

        for cls in mod.classes:
            _index_class(index, cls, mod.name, mod_route)

        for func in mod.functions:
            fqn = f"{mod.name}.{func.name}"
            index[fqn] = f"{mod_route}#{func.name.lower()}"

    return index


def _index_class(
    index: dict[str, str],
    cls: ClassIR,
    parent_fqn: str,
    mod_route: str,
) -> None:
    """Index a class and its inner classes recursively."""
    fqn = f"{parent_fqn}.{cls.name}"
    index[fqn] = f"{mod_route}#{cls.name.lower()}"

    for inner in cls.inner_classes:
        _index_class(index, inner, fqn, mod_route)


def _extract_bare_names(type_str: str) -> list[str]:
    """Extract all potential symbol names from a type string.

    Handles generics like ``list[Config]``, unions like ``Config | None``,
    and comma-separated types like ``dict[str, Config]``.
    """
    # Split union types (X | Y)
    parts = _UNION_PIPE_RE.split(type_str)

    names: list[str] = []
    for part in parts:
        part = part.strip()
        if not part:
            continue
        m = _GENERIC_RE.match(part)
        if m:
            # Recurse into the generic's inner types
            inner = m.group(2)
            # Split on commas but respect nested brackets
            inner_parts = _split_respecting_brackets(inner)
            for ip in inner_parts:
                names.extend(_extract_bare_names(ip.strip()))
        else:
            names.append(part)
    return names


def _split_respecting_brackets(s: str) -> list[str]:
    """Split a string on commas, but not inside brackets."""
    parts: list[str] = []
    depth = 0
    current: list[str] = []
    for ch in s:
        if ch in "([":
            depth += 1
            current.append(ch)
        elif ch in ")]":
            depth -= 1
            current.append(ch)
        elif ch == "," and depth == 0:
            parts.append("".join(current))
            current = []
        else:
            current.append(ch)
    if current:
        parts.append("".join(current))
    return parts


def resolve_type_link(
    type_str: str,
    index: dict[str, str],
    current_module: str,
) -> str | None:
    """Try to resolve a type string to a documentation URL.

    Handles:
    - Simple names: ``Config`` -> look up ``{current_module}.Config``, then all modules
    - Qualified names: ``folio.config.Config`` -> direct lookup
    - Generic types: ``list[Config]`` -> resolve the inner type ``Config``
    - Union types: ``Config | None`` -> resolve ``Config``

    Returns the URL if a single resolvable symbol is found, None otherwise.
    """
    type_str = type_str.strip()
    if not type_str:
        return None

    # Try direct lookup first (fully-qualified name)
    if type_str in index:
        return index[type_str]

    # Extract all bare names from the type expression
    bare_names = _extract_bare_names(type_str)

    # Filter out builtins and empty strings
    candidates = [
        n for n in bare_names if n and n not in _BUILTINS and not n.startswith("_")
    ]

    if len(candidates) != 1:
        # Ambiguous or no candidates -- don't link
        return None

    name = candidates[0]

    # Direct lookup (fully qualified)
    if name in index:
        return index[name]

    # Try current_module.name
    local_fqn = f"{current_module}.{name}"
    if local_fqn in index:
        return index[local_fqn]

    # Try parent package.name (e.g., current_module=folio.config, try folio.name)
    if "." in current_module:
        parent = current_module.rsplit(".", 1)[0]
        parent_fqn = f"{parent}.{name}"
        if parent_fqn in index:
            return index[parent_fqn]

    # Search all entries for a matching suffix
    suffix = f".{name}"
    matches = [url for fqn, url in index.items() if fqn.endswith(suffix)]
    if len(matches) == 1:
        return matches[0]

    return None

File: crates/folio-docs/src/xref.rs
//! Cross-references: the project-wide symbol index (`fqn -> URL`) and the
//! resolution of a type string to one link.

use std::collections::BTreeMap;

use folio_ir::{ClassIR, ModuleIR, TypeKind};

use crate::routes::{
    anchor_slug, member_anchor, module_route, route_base, symbol_prefix, type_anchor,
};

/// Fully-qualified symbol -> site-absolute URL (plus `#anchor`). A `BTreeMap`
/// so its compact JSON is canonical: the manifest hashes it as `symbols`.
pub type SymbolIndex = BTreeMap<String, String>;

/// Python builtins and typing names, and JavaScript's `null` and `undefined`,
/// that never link.
pub const BUILTINS: &[&str] = &[
    "str",
    "int",
    "float",
    "bool",
    "bytes",
    "bytearray",
    "list",
    "dict",
    "set",
    "frozenset",
    "tuple",
    "None",
    "type",
    "object",
    "complex",
    "range",
    "memoryview",
    "slice",
    "property",
    "classmethod",
    "staticmethod",
    "super",
    "Exception",
    "BaseException",
    "Any",
    "Callable",
    "Iterator",
    "Generator",
    "Coroutine",
    "Awaitable",
    "AsyncIterator",
    "AsyncGenerator",
    "Iterable",
    "Sequence",
    "Mapping",
    "MutableMapping",
    "MutableSequence",
    "MutableSet",
    "Optional",
    "Union",
    "Type",
    "ClassVar",
    "Final",
    "Literal",
    "Protocol",
    "TypeVar",
    "TypeAlias",
    "Self",
    "Never",
    "NoReturn",
    "Concatenate",
    "ParamSpec",
    "TypeVarTuple",
    "Unpack",
    "null",
    "undefined",
];

/// Index every module, class (inner classes as `Outer.Inner`), function and
/// type under `docs_route_base` (`/docs` when blank), each at the id its page
/// writes. An `impl` block is not a symbol: it never takes its type's entry.
/// Later duplicates win.
pub fn build_symbol_index(modules: &[ModuleIR], docs_route_base: &str) -> SymbolIndex {
    let base = route_base(docs_route_base);
    let mut index = SymbolIndex::new();
    for module in modules {
        let mod_route = format!("{base}/{}", module_route(module));
        let prefix = symbol_prefix(module);
        index.insert(prefix.clone(), mod_route.clone());
        for class in &module.classes {
            index_class(&mut index, class, &prefix, &mod_route);
        }
        for func in &module.functions {
            index.insert(
                format!("{prefix}.{}", func.name),
                format!("{mod_route}#{}", anchor_slug(&func.name)),
            );
        }
        for item in module.types.iter().filter(|t| t.kind != TypeKind::Impl) {
            index.insert(
                format!("{prefix}.{}", item.name),
                format!("{mod_route}#{}", type_anchor(item)),
            );
        }
    }
    index
}

fn index_class(index: &mut SymbolIndex, class: &ClassIR, parent: &str, mod_route: &str) {
    index_class_at(index, class, parent, mod_route, &anchor_slug(&class.name));
}

/// An inner class sits at `outer-inner`, the id its `ClassOverview` carries.
fn index_class_at(
    index: &mut SymbolIndex,
    class: &ClassIR,
    parent: &str,
    mod_route: &str,
    anchor: &str,
) {
    let fqn = format!("{parent}.{}", class.name);
    index.insert(fqn.clone(), format!("{mod_route}#{anchor}"));
    for inner in &class.inner_classes {
        let inner_anchor = member_anchor(anchor, &inner.name);
        index_class_at(index, inner, &fqn, mod_route, &inner_anchor);
    }
}

/// `head[args]` or `head<args>` with a dotted head (`typing.Optional[Config]`,
/// `Promise<Connection>`, JSDoc's `Array.<Connection>`): the head and the
/// arguments' text.
fn generic_parts(part: &str) -> Option<(&str, &str)> {
    let open = part.find(['[', '<'])?;
    let (head, close) = match part[..open].trim_end() {
        head if part[open..].starts_with('[') => (head, ']'),
        head => (head.strip_suffix('.').unwrap_or(head), '>'),
    };
    let inner = part.strip_suffix(close)?.get(open + 1..)?;
    let is_name = |segment: &str| {
        !segment.is_empty() && segment.chars().all(|c| c.is_alphanumeric() || c == '_')
    };
    (head.split('.').all(is_name) && !inner.trim().is_empty()).then_some((head, inner))
}

/// The text inside one pair of matching quotes (`'Config'`, `"Config"`), or
/// the text as it is.
fn unquote(s: &str) -> &str {
    ['\'', '"']
        .into_iter()
        .find_map(|q| s.strip_prefix(q)?.strip_suffix(q))
        .map_or(s, str::trim)
}

/// A name without JSDoc's leading `?`/`!` or an array's trailing `[]`
/// (`?Connection`, `Connection[]` -> `Connection`).
fn strip_modifiers(mut s: &str) -> &str {
    s = s.trim_start_matches(['?', '!']).trim_start();
    while let Some(rest) = s.strip_suffix("[]") {
        s = rest.trim_end();
    }
    s
}

/// Every candidate name in a type string: union members, and the arguments of
/// generics recursively (`dict[str, Config]` -> `str`, `Config`). Quotes, a
/// leading `?`/`!` and a trailing `[]` come off each name, so `'Config'`,
/// `?Config` and `Config[]` all give `Config`. A `Literal[...]` holds values,
/// not names.
pub fn extract_bare_names(type_str: &str) -> Vec<String> {
    let mut names = Vec::new();
    for part in split_outside_brackets(unquote(type_str.trim()), '|') {
        let part = strip_modifiers(unquote(part.trim()));
        if part.is_empty() {
            continue;
        }
        match generic_parts(part) {
            Some((head, _)) if head.rsplit('.').next() == Some("Literal") => {}
            Some((_, inner)) => {
                for item in split_respecting_brackets(inner) {
                    names.extend(extract_bare_names(item.trim()));
                }
            }
            None => names.push(part.to_string()),
        }
    }
    names
}

/// Split on commas outside `()`/`[]`/`<>`.
pub fn split_respecting_brackets(s: &str) -> Vec<String> {
    split_outside_brackets(s, ',')
}

/// Split on `sep` outside `()`/`[]`/`<>`; the `>` of `->` or `=>` closes nothing.
fn split_outside_brackets(s: &str, sep: char) -> Vec<String> {
    let mut parts = Vec::new();
    let mut current = String::new();
    let mut depth = 0i32;
    let mut prev = '\0';
    for ch in s.chars() {
        match ch {
            '(' | '[' | '<' => depth += 1,
            '>' if prev == '-' || prev == '=' => {}
            ')' | ']' | '>' => depth -= 1,
            _ if ch == sep && depth == 0 => {
                parts.push(std::mem::take(&mut current));
                prev = ch;
                continue;
            }
            _ => {}
        }
        current.push(ch);
        prev = ch;
    }
    if !current.is_empty() {
        parts.push(current);
    }
    parts
}

/// The language of an index key: `""` for Python, which has no `language:`
/// prefix, and `rust` for `rust:a::b`, whose `::` comes after the prefix.
fn key_language(key: &str) -> &str {
    key.split_once(':').map_or("", |(language, _)| language)
}

/// The URL a type string links to, when exactly one non-builtin name in it
/// resolves: direct key, `current_module.name`, the parent package, then a
/// unique `.name` suffix among the keys of the current module's language.
pub fn resolve_type_link(
    type_str: &str,
    index: &SymbolIndex,
    current_module: &str,
) -> Option<String> {
    let type_str = type_str.trim();
    if type_str.is_empty() {
        return None;
    }
    if let Some(url) = index.get(type_str) {
        return Some(url.clone());
    }
    let names = extract_bare_names(type_str);
    let mut candidates = names
        .iter()
        .filter(|n| !BUILTINS.contains(&n.as_str()) && !n.starts_with('_'));
    let name = candidates.next()?;
    if candidates.next().is_some() {
        return None;
    }
    if let Some(url) = index.get(name) {
        return Some(url.clone());
    }
    if let Some(url) = index.get(&format!("{current_module}.{name}")) {
        return Some(url.clone());
    }
    if let Some((parent, _)) = current_module.rsplit_once('.') {
        if let Some(url) = index.get(&format!("{parent}.{name}")) {
            return Some(url.clone());
        }
    }
    let suffix = format!(".{name}");
    let language = key_language(current_module);
    let mut matches = index
        .iter()
        .filter(|(fqn, _)| fqn.ends_with(&suffix) && key_language(fqn) == language)
        .map(|(_, url)| url);
    let first = matches.next()?;
    matches.next().is_none().then(|| first.clone())
}

#[cfg(test)]
#[path = "xref_tests.rs"]
mod tests;
